pkg/build/builder/common.go

```diff
@@ -82,3 +91,141 @@ func updateBuildRevision(c client.BuildInterface, build *api.Build, sourceInfo *
 		glog.Warningf("An error occurred saving build revision: %v", err)
 	}
 }
+
+// randomBuildTag generates a random tag used for building images in such a way
+// that the built image can be referred to unambiguously even in the face of
+// concurrent builds with the same name in the same namespace.
+func randomBuildTag(namespace, name string) string {
+	repo := fmt.Sprintf("%s/%s", namespace, name)
+	randomTag := fmt.Sprintf("%08x", rand.Uint32())
+	maxRepoLen := reference.NameTotalLengthMax - len(randomTag) - 1
+	if len(repo) > maxRepoLen {
+		repo = fmt.Sprintf("%x", sha1.Sum([]byte(repo)))
+	}
+	return fmt.Sprintf("%s:%s", repo, randomTag)
+}
+
+// containerNamePrefix prefixes the name of containers launched by a build. We
+// cannot reuse the prefix "k8s" because we don't want the containers to be
+// managed by a kubelet.
+const containerNamePrefix = "openshift"
+
+// containerName creates names for Docker containers launched by a build. It is
+// meant to resemble Kubernetes' pkg/kubelet/dockertools.BuildDockerName.
```

Comment:
There should never be two concurrent build "names" - that would be an error case.  If build name here is the Build object name, then having two containers for that same build would be an error and we should be cleaned up prior to the build, or cause the build to fail.  The next build would have a different build name so it would not fail.

The random suffix means that we don't actually know what container our previous build runs with.  We generally don't want to use random names anywhere except at the top most level of the system (if a user actually says "hey don't care about this") - the rest of the system needs to be using predictable names all the way down, because that guarantees consistency.

In general, use of random anywhere is a code smell in a distributed system, because it means two components can't agree on the same way to reference something.